Implement more linux/amd64 syscalls

samhocevar opened the pull request.

samhocevar committed:
Replace syscall SYS_clock_gettime with walltime1() on linux/amd64

samhocevar committed:
Add empty C stubs for the fcntl, fstat, lseek, renameat, and fstatat functions

samhocevar committed:
Hook C stub syscalls into the Go runtime

samhocevar committed:
Implement openat, mkdirat, and getdents64 syscalls

We actually replace openat() with a call to open() in runtime.Open()

samhocevar committed:
Disable the TestMirrorWithReflect test because it requires parsing directories

hajimehoshi commented:
Could you tell me more context about this PR?

hajimehoshi commented on 1.19_linux/runtime/sys_linux_amd64.s.patch:
```diff
@@ -347,65 +347,65 @@ TEXT runtime·sysUnusedOS_trampoline(SB),NOSPLIT,$0
+	CALL	hitsumabushi_sysMapOS(SB)
 	POPQ	BP
-        RET
+	RET
```
Optional: indentation fix can be in another PR

hajimehoshi commented:
Friendly ping

samhocevar commented:
Hi! I’m sorry I’m still a bit struggling with a nasty Covid, but I’ll try to be clear.

This PR (with a few additional minor fixes) allows building the Go runtime with the native PS4 and PS5 compilers (also the Switch compiler). It also moves the special logic for file descriptors 0/1/2 to a single place.

The idea is then to let the user reimplement `c_write()`, `c_fnctl()` etc. with their implementation of choice.

Right now I can run a simple Lua-based application on the PlaysStation.

I understand that makes a lot of syscalls handled in `syscall_linux.go`. Maybe you’d prefer this to remain a very simple wrapper and have it implemented as a couple of large syscall/syscall6 C functions that the user can override?

hajimehoshi commented on 1.19_linux/runtime/sys_linux_arm64.s.patch:
```diff
@@ -503,3 +503,49 @@ TEXT runtime·sysMapOS_trampoline(SB),NOSPLIT,$0
+	MOVW	R0, 32(R19)	// return value
+	RET
+
+TEXT runtime·fstatat_trampoline(SB),NOSPLIT,$0
```
Let's order the functions in a consistent way with _amd64 version.

samhocevar replied:
Done!

hajimehoshi commented on 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c:
```diff
@@ -265,12 +267,47 @@ int32_t c_write1(uintptr_t fd, void *p, int32_t n) {
     break;
   default:
     fprintf(stderr, "syscall write(%lu, %p, %d) is not implemented\n", fd, p, n);
+    ret = -ENOSYS;
```
EBADF?

https://man7.org/linux/man-pages/man2/write.2.html

samhocevar replied:
Yes, no strong opinion but it makes sense.

hajimehoshi commented on 1.19_linux/runtime/cgo/hitsumabushi_syscalls_linux.c:
```diff
@@ -265,12 +267,47 @@ int32_t c_write1(uintptr_t fd, void *p, int32_t n) {
+    }
+  }
+  fprintf(stderr, "syscall fcntl(%d, %d, %d) is not implemented\n", fd, cmd, arg);
+  return -ENOSYS;
```
EBADF?

samhocevar replied:
Done!

hajimehoshi commented:
> The idea is then to let the user reimplement c_write(), c_fnctl() etc. with their implementation of choice.

I see. So we need more changes to add an API to provide the filename to overwrite later, right?

samhocevar commented:
> > The idea is then to let the user reimplement c_write(), c_fnctl() etc. with their implementation of choice.
> 
> I see. So we need more changes to add an API to provide the filename to overwrite later, right?

Yes, that would be the next step. I just don’t know about the granularity, because maybe not everything should go to the same place. All filesystem functions could go to the same file, but for instance `c_sched_getaffinity()` is maybe more related to `c_futex()`, but I’m not sure about `c_walltime1()`.

samhocevar committed:
Use EBADF in system call stubs when more appropriate than ENOSYS

samhocevar commented:
Should the functions be named `hitsumabushi_open()` etc. instead of `c_open()` etc.?

hajimehoshi commented:
> Should the functions be named hitsumabushi_open() etc. instead of c_open() etc.?

Yes I think so.

samhocevar commented:
So, this is what I would like to suggest:
 - move the following functions to `hitsumabushi_clock_linux.c`:
   - `hitsumabushi_nanotime1()`
   - `hitsumabushi_usleep()`
   - `hitsumabushi_walltime1()`
 - move the following functions to `hitsumabushi_cpu_linux.c`:
   - `hitsumabushi_sched_getaffinity()`
 - create a new `hitsumabushi_filesystem_linux.c` or `hitsumabushi_fs_linux.c` and move the following functions there:
   - `hitsumabushi_open()`
   - `hitsumabushi_read()`
   - `hitsumabushi_write1()`
   - `hitsumabushi_lseek()`
   - `hitsumabushi_fcntl()`
   - `hitsumabushi_fstat()`
   - `hitsumabushi_renameat()`
   - `hitsumabushi_fstatat()`
   - `hitsumabushi_closefd()`

This will only leave the following functions in `hitsumabushi_syscalls_linux.c`:
 - `hitsumabushi_exit()`
 - `hitsumabushi_gettid()`
 - `hitsumabushi_osyield()`

These functions are very threading- and process-related, maybe they could be merged together with the `futex` functions into something like `hitshumabushi_thread_linux.c`? Just in case we want to limit the number of extra C files.

Also, the `1` suffix in `c_write1()` is only here because there was already a `write()` function in the Go runtime. I would like to suggest dropping the suffix and go for `hitsumabushi_write()` (same for the other similar function), because it’s really just an implementation detail that the user doesn’t really need to know about.

hajimehoshi commented:
> These functions are very threading- and process-related, maybe they could be merged together with the futex functions into something like hitshumabushi_thread_linux.c? Just in case we want to limit the number of extra C files.

Sounds a good idea! This would break compatibility but that's fine as this is still v0.

> Also, the 1 suffix in c_write1() is only here because there was already a write() function in the Go runtime. I would like to suggest dropping the suffix and go for hitsumabushi_write() (same for the other similar function), because it’s really just an implementation detail that the user doesn’t really need to know about.

write1 is a very special function to dump a message from panic. So I recommend keeping 1. In general, I prefer keeping the Go function names whenever possible.

https://cs.opensource.google/go/go/+/master:src/runtime/os_windows.go;l=706-711;drc=e252dcf9d38ce9192bccacb7e33867cbfbd22b6c

```
// write1 must be nosplit because it's used as a last resort in
// functions like badmorestackg0. In such cases, we'll always take the
// ASCII path.
```

samhocevar committed:
Rename hitsumabushi_futex_linux.c to hitsumabushi_threads_linux.c

This file will eventually contain more threading-related functions.

ReplaceFutex() is unchanged because it is legacy, but FutexFilePath() is now
called ThreadsFilePath().

samhocevar committed:
Move threading-related functions to hitsumabushi_threads_linux.c

The functions are: c_gettid(), c_osyield(), and c_exit(). They are now prefixed
with “hitsumabushi_”.

samhocevar committed:
Move clock-related functions to hitsumabushi_clock_linux.c

The functions are c_usleep(), c_walltime1(), and c_nanotime1(). They are now
prefixed with “hitsumabushi_”.

As a result the function hitsumabushi_clock_gettime() was removed because it
was just an implementation helper for walltime1() and nanotime1().

samhocevar committed:
Move c_sched_getaffinity to hitsumabushi_cpu_linux.c

samhocevar committed:
Move filesystem-related files to hitsumabushi_filesystem_linux.c

There is also a new FilesystemFilePath() function to override this file.

samhocevar committed:
Remove hitsumabushi_syscalls_linux.c

This file now only contains unused functions.

samhocevar commented:
I think the work is done! Now `overlay.go` has the following replacement functions:
 - `func ClockFilePath(os string) (string, error)`
 - `func ThreadsFilePath(os string) (string, error)`
 - `func FilesystemFilePath(os string) (string, error)`
 - `func MemoryFilePath(os string) (string, error)`
 - `func CPUFilePath(os string) (string, error)`

Maybe `ThreadsFilePath` and `CPUFilePath` could be merged even more? The rest looks quite well grouped by theme.

Also I noticed in https://github.com/hajimehoshi/hitsumabushi/pull/13/commits/1ca1ec100fcb38f4a0cf79a39a868b74d8d35585 that a lot of functions were actually not used, does that make sense to you?

hajimehoshi commented on 1.19_linux/runtime/cgo/hitsumabushi_clock_linux.c:
```diff
@@ -1,8 +1,23 @@
+  struct timespec tp;
+  clock_gettime(CLOCK_MONOTONIC, &tp);
+  return (int64_t)(tp.tv_sec) * 1000000000ll + (int64_t)tp.tv_nsec;
 }
```
This will require Nintendo Switch side, right? Did you need these functions ?

samhocevar replied:
I am not sure I understand the question. The above works fine for me on Switch. Could that be because I use `Compilers/NX/nx/aarch64/bin/clang.exe` but you use a Linux compiler?

hajimehoshi replied:
Let's talk in the Discord.

hajimehoshi commented on 1.19_linux/runtime/cgo/hitsumabushi_cpu_linux.c:
```diff
@@ -1,8 +1,20 @@
 	return 1;
 }
+
+int32_t hitsumabushi_sched_getaffinity(pid_t pid, size_t cpusetsize, void *mask) {
```
Do we really need to expose this for rewriting?

samhocevar replied:
Probably not required, indeed. The Sony consoles have an equivalent for this function so I was happy to use it, but it’s really not very important.

hajimehoshi commented on 1.19_linux/runtime/time_linux_amd64.s.patch:
```diff
@@ -0,0 +1,4 @@
+//--from
+//go:build !faketime
+//--to
+//go:build !faketime && false
```
Just `//go:build ignore`?

samhocevar replied:
OK!

hajimehoshi reviewed (commented):
I'm a little skeptical about exposing some functions as commented. Did you actually need to overwrite these functions on your platform?

hajimehoshi commented:
Yes, mmap and signals are not available on Nintendo switch.

2022年11月18日(金) 21:03 Sam Hocevar ***@***.***>:

> I think the work is done! Now overlay.go has the following replacement
> functions:
>
>    - func ClockFilePath(os string) (string, error)
>    - func ThreadsFilePath(os string) (string, error)
>    - func FilesystemFilePath(os string) (string, error)
>    - func MemoryFilePath(os string) (string, error)
>    - func CPUFilePath(os string) (string, error)
>
> Maybe ThreadsFilePath and CPUFilePath could be merged even more? The rest
> looks quite well grouped by theme.
>
> Also I noticed in 1ca1ec1
> <https://github.com/hajimehoshi/hitsumabushi/commit/1ca1ec100fcb38f4a0cf79a39a868b74d8d35585>
> that a lot of functions were actually not used, does that make sense to you?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hajimehoshi/hitsumabushi/pull/13#issuecomment-1319908473>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEENVJXY2JDWHS7BGKCNLWI5V77ANCNFSM6AAAAAARRK3TVI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

samhocevar commented:
> Yes, mmap and signals are not available on Nintendo switch.

I would say that they belong to a file outside of Hitsumabushi, then. I think you need them because your compiler generates dependencies to `mmap` etc., but when using the Nintendo or Switch compilers, these functions are not needed, and at least on the PlayStation they cause conflicts with the system’s libraries.

hajimehoshi commented on 1.19_linux/runtime/sys_linux_amd64.s.patch:
```diff
@@ -21,7 +21,7 @@ TEXT runtime·open_trampoline(SB),NOSPLIT,$0
 	MOVL	12(DI), DX		// arg 3 mode
 	MOVQ	0(DI), DI		// arg 1 pathname
 	XORL	AX, AX			// vararg: say "no float args"
-	CALL	c_open(SB)
```
I'm afraid this might be a burden on you, but I think it'd be nice to have a separate PR to rename the existing `c_` functions to `hitsumabushi_` functions

samhocevar committed:
1.19_linux/runtime: remove calls to sigfillset() and pthread_sigmask()

These functions are stubs in Hitsumabushi, so no need to call them.

hajimehoshi commented on 1.19_linux/runtime/cgo/gcc_linux_arm64.c.patch:
```diff
@@ -28,23 +28,17 @@ void
 	ts->g->stackhi = size;
 	err = _cgo_try_pthread_create(&p, &attr, threadentry, ts);
 
-	pthread_sigmask(SIG_SETMASK, &oset, nil);
```
Removing pthread functions can be a pure refactoring in an independent PR.

samhocevar commented:
I think this PR can be abandoned. It has been split into many smaller PRs (#17, #19, #20) and the remaining differences are actually not very important because there is no use case for overriding all these other functions (at least for now).

hajimehoshi commented:
Sure, let's revisit this when we find more functions we want to overwrite.
